debian/mongooplog.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOOPLOG" "1" "January 30, 2015" "3.0" "mongodb-manual"
.SH NAME
mongooplog \- MongoDB Oplog Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.sp
New in version 2.2.

.SH SYNOPSIS
.sp
\fBmongooplog\fP is a simple tool that polls operations from
the \fIreplication\fP \fIoplog\fP of a remote server, and applies
them to the local server. This capability supports certain classes of
real\-time migrations that require that the source server remain online
and in operation throughout the migration process.
.sp
Typically this command will take the following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongooplog  \-\-from mongodb0.example.net \-\-host mongodb1.example.net
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
This command copies oplog entries from the \fBmongod\fP instance
running on the host \fBmongodb0.example.net\fP and duplicates
operations to the host \fBmongodb1.example.net\fP\&. If you do not need
to keep the \fI\-\-from\fP host running during
the migration, consider using \fBmongodump\fP and
\fBmongorestore\fP or another \fBbackup\fP operation, which may be better suited to
your operation.
.sp
\fBNOTE:\fP
.INDENT 0.0
.INDENT 3.5
If the \fBmongod\fP instance specified by the \fI\-\-from\fP
argument is running with \fBauthentication\fP, then
\fBmongooplog\fP will not be able to copy oplog entries.
.UNINDENT
.UNINDENT
.sp
\fBSEE ALSO:\fP
.INDENT 0.0
.INDENT 3.5
\fBmongodump\fP, \fBmongorestore\fP,
http://docs.mongodb.org/manual/core/backups, http://docs.mongodb.org/manual/core/replica\-set\-oplog\&.
.UNINDENT
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fBmongooplog\fP removed the \fB\-\-dbpath\fP as well as related
\fB\-\-directoryperdb\fP and \fB\-\-journal\fP options. You must use
\fBmongooplog\fP while connected to a \fBmongod\fP instance.

.INDENT 0.0
.TP
.B mongooplog
.UNINDENT
.INDENT 0.0
.TP
.B mongooplog
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongooplog\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs the \fBmongooplog\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongooplog\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
Specifies a resolvable hostname for the \fBmongod\fP instance to
which \fBmongooplog\fP will apply \fIoplog\fP operations
retrieved from the server specified by the \fI\-\-from\fP option.
.sp
By default \fBmongooplog\fP attempts to connect to a MongoDB instance running
on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the \fBreplica set name\fP and a seed list of set members. Use the
following form:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single MongoDB instance by
specifying the host and port number directly.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-port
Specifies the port number of the \fBmongod\fP instance where
\fBmongooplog\fP will apply \fIoplog\fP entries. Specify
this option only if the MongoDB instance to connect to is not
running on the standard port of \fB27017\fP\&. You may also specify a
port number using the \fI\-\-host\fP command.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
Enables IPv6 support and allows the \fBmongooplog\fP to connect to the
MongoDB instance using an IPv6 network. All MongoDB programs and
processes disable IPv6 support by default.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongod\fP or \fBmongos\fP that has
SSL support enabled.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
If the \fBmongo\fP shell or any other tool that connects to
\fBmongos\fP or \fBmongod\fP is run without
\fI\-\-sslCAFile\fP, it will not attempt to validate
server certificates. This results in vulnerability to expired
\fBmongod\fP and \fBmongos\fP certificates as well as to foreign
processes posing as valid \fBmongod\fP or \fBmongos\fP
instances. Ensure that you \fIalways\fP specify the CA file against which
server certificates should be validated in cases where intrusion is a
possibility.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\-\-ssl\fP option to connect
to a \fBmongod\fP or \fBmongos\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp
Specifies the password to de\-crypt the certificate\-key file (i.e.
\fI\-\-sslPEMKeyFile\fP). Use the \fI\-\-sslPEMKeyPassword\fP option only if the
certificate\-key file is encrypted. In all cases, the \fBmongooplog\fP will
redact the password from all logging and reporting output.
.sp
If the private key in the PEM file is encrypted and you do not specify
the \fI\-\-sslPEMKeyPassword\fP option, the \fBmongooplog\fP will prompt for a passphrase. See
\fIssl\-certificate\-password\fP\&.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCRLFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the Certificate Revocation
List. Specify the file name of the \fB\&.pem\fP file using relative or
absolute paths.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidCertificates
New in version 2.6.

.sp
Bypasses the validation checks for server certificates and allows
the use of invalid certificates. When using the
\fBallowInvalidCertificates\fP setting, MongoDB logs as a
warning the use of the invalid certificate.
.sp
The default distribution of MongoDB does not contain support for SSL.
For more information on MongoDB and SSL, see http://docs.mongodb.org/manual/tutorial/configure\-ssl\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslAllowInvalidHostnames
New in version 3.0.

.sp
Disables the validation of the hostnames in SSL certificates. Allows
\fBmongooplog\fP to connect to MongoDB instances if the hostname their
certificates do not match the specified hostname.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslFIPSMode
New in version 2.6.

.sp
Directs the \fBmongooplog\fP to use the FIPS mode of the installed OpenSSL
library. Your system must have a FIPS compliant OpenSSL library to use
the \fI\-\-sslFIPSMode\fP option.
.sp
\fBNOTE:\fP
.INDENT 7.0
.INDENT 3.5
FIPS Compatible SSL is
available only in \fI\%MongoDB Enterprise\fP\&. See
http://docs.mongodb.org/manual/tutorial/configure\-fips for more information.
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-username <username>, \-u <username>
Specifies a username with which to authenticate to a MongoDB database
that uses authentication. Use in conjunction with the \fB\-\-password\fP and
\fB\-\-authenticationDatabase\fP options.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-password <password>, \-p <password>
Specifies a password with which to authenticate to a MongoDB database
that uses authentication. Use in conjunction with the \fB\-\-username\fP and
\fB\-\-authenticationDatabase\fP options.
.sp
If you do not specify an argument for \fI\-\-password\fP, \fBmongooplog\fP will
prompt interactively for a password on the console.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationDatabase <dbname>
New in version 2.4.

.sp
Specifies the database that holds the user\(aqs credentials.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-authenticationMechanism <name>
\fIDefault\fP: MONGODB\-CR
.sp
New in version 2.4.

.sp
Changed in version 2.6: Added support for the \fBPLAIN\fP and \fBMONGODB\-X509\fP authentication
mechanisms.

.sp
Specifies the authentication mechanism the \fBmongooplog\fP instance uses to
authenticate to the \fBmongod\fP or \fBmongos\fP\&.
.TS
center;
|l|l|.
_
T{
Value
T}	T{
Description
T}
_
T{
MONGODB\-CR
T}	T{
MongoDB challenge/response authentication.
T}
_
T{
MONGODB\-X509
T}	T{
MongoDB SSL certificate authentication.
T}
_
T{
PLAIN
T}	T{
External authentication using LDAP. You can also use \fBPLAIN\fP
for authenticating in\-database users. \fBPLAIN\fP transmits
passwords in plain text. This mechanism is available only in
\fI\%MongoDB Enterprise\fP\&.
T}
_
T{
GSSAPI
T}	T{
External authentication using Kerberos. This mechanism is
available only in \fI\%MongoDB Enterprise\fP\&.
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiServiceName
New in version 2.6.

.sp
Specify the name of the service using \fBGSSAPI/Kerberos\fP\&. Only required if the service does not use the
default name of \fBmongodb\fP\&.
.sp
This option is available only in MongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-gssapiHostName
New in version 2.6.

.sp
Specify the hostname of a service using \fBGSSAPI/Kerberos\fP\&. \fIOnly\fP required if the hostname of a machine does
not match the hostname resolved by DNS.
.sp
This option is available only in MongoDB Enterprise.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-db <database>, \-d <database>
Specifies the name of the database on which to run the \fBmongooplog\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-collection <collection>, \-c <collection>
Specifies the collection to export.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-seconds <number>, \-s <number>
Specify a number of seconds of operations for \fBmongooplog\fP to
pull from the \fIremote host\fP\&. Unless
specified the default value is \fB86400\fP seconds, or 24 hours.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-from <host[:port]>
Specify the host for \fBmongooplog\fP to retrieve \fIoplog\fP
operations from. \fBmongooplog\fP \fIrequires\fP this option.
.sp
Unless you specify the \fI\-\-host\fP option,
\fBmongooplog\fP will apply the operations collected with this
option to the oplog of the \fBmongod\fP instance running on the
localhost interface connected to port \fB27017\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-oplogns <namespace>
Specify a namespace in the \fI\-\-from\fP host
where the oplog resides. The default value is \fBlocal.oplog.rs\fP, which
is the where \fIreplica set\fP members store their operation log.
However, if you\(aqve copied \fIoplog\fP entries into another
database or collection or are pulling oplog entries from a
\fBmaster\-slave\fP deployment, use \fI\%\-\-oplogns\fP to
apply oplog entries stored in another location. Namespaces take the
form of \fB[database].[collection]\fP\&.
.UNINDENT
.SH USE
.sp
Consider the following prototype \fBmongooplog\fP command:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongooplog  \-\-from mongodb0.example.net \-\-host mongodb1.example.net
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Here, entries from the \fIoplog\fP of the \fBmongod\fP running
on port \fB27017\fP\&. This only pull entries from the last 24 hours.
.sp
Use the \fI\-\-seconds\fP argument to capture
a greater or smaller amount of time. Consider the following example:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongooplog  \-\-from mongodb0.example.net \-\-seconds 172800
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
In this operation, \fBmongooplog\fP captures 2 full days of
operations. To migrate 12 hours of \fIoplog\fP entries, use the
following form:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongooplog  \-\-from mongodb0.example.net \-\-seconds 43200
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
MongoDB Documentation Project
.SH COPYRIGHT
2011-2015
.\" Generated by docutils manpage writer.
.


NEWS
WiredTiger release 2.8.0, 2015-22-03
------------------------------------

The WiredTiger 2.8.0 release contains new features, new supported platforms,
minor API changes and bug fixes.

New features and API changes; refer to the API documentation for full details:

* WT-60         Port WiredTiger to run on big endian platforms
* WT-2287       Add a new WT_SESSION.rebalance API
* WT-2333       Add a lock_wait configuration setting to WT_SESSION.drop to avoid blocking
* WT-2349       Add a readonly configuration setting to wiredtiger_open
* WT-2363       Remove built in support for bzip2 compression
* WT-2404       Add streaming pack/unpack methods to the extension API

Significant changes and bug fixes:

* WT-1801       Add a directory sync after rollback of a WT_SESSION::rename operation
* WT-2130       Improve on-disk page utilization with random workloads
* WT-2275       Fix a database corruption after truncate and crash
* WT-2264       High update workloads can cause checkpoints to never complete
* WT-2290       Improve effectiveness of WT_SESSION.compact
* WT-2361       Fix a bug in column-store where verify identifies out of order data
* WT-2367       Fix a bug in WT_CURSOR.next that could cause out-of-order key returns
* WT-2374       Fix a bug where a database was corrupted when restoring a backup
* WT-2381       Fix the dump utility to include the table configuration
* WT-2451       Allow the WiredTiger metadata to be evicted
* WT-2490       Fix a bug in column-store where search_near() returns the wrong key

Issues fixed in MongoDB:

* SERVER-21619  sys-perf: WT crash during core_workloads_WT execution
* SERVER-21833  Enhance WT_SESSION::compact to more reliably release space
* SERVER-21887  Enhance $sample to be faster on newly created collection
* SERVER-22676  Allow WiredTiger to open databases created by 3.0.0 or 3.0.1
* SERVER-22773  New CRC32 implementation on PowerPC
* SERVER-22831  Low query rate with heavy cache pressure and an idle collection

Other noteworthy changes since the previous release:

* WT-1517       Fix error handling around schema format edge cases
* WT-2060       Simplify aggregation of statistics
* WT-2073       Metadata cleanups
* WT-2099       Seeing memory underflow messages
* WT-2107       Add example code including an event handler
* WT-2113       Truncate test occasionally fails with unexpected EBUSY
* WT-2123       Don't clear allocated memory if not required
* WT-2173       Fix some cases where tiny caches could get stuck full
* WT-2177       Add an optional per-thread seed to random number generator
* WT-2198       Bulk load and column store appends
* WT-2215       WT_LSN needs to support atomic reads and updates
* WT-2216       Simplify row-store search loop slightly
* WT-2231       Pinned page cursor searches could check parent keys
* WT-2235       Add a unicode option to WiredTiger printlog utility
* WT-2242       WiredTiger treats dead trees the same as other trees in eviction
* WT-2246       Improve performance for column-store append searches
* WT-2247       Variable-length column-store in-memory page splits
* WT-2258       Stop WiredTiger pre-loading pages when direct-IO is configured
* WT-2259       Fix error handling when getting exclusive access to a btree
* WT-2262       Fix random cursor next so it is not skewed by tree shape
* WT-2265       WiredTiger related change in PowerPC specific code block in gcc.h
* WT-2272       Fix a bug in the sweep server that triggered an assertion
* WT-2276       Add a tool to decode checkpoint addr
* WT-2277       Remove WT check against big-endian systems
* WT-2279       Define WT_PAUSE(), WT_FULL_BARRIER(), etc when s390x is defined
* WT-2280       Add CRC32 Optimized code for PowerPC
* WT-2282       Error in wt_txn_update_oldest verbose message test
* WT-2283       Retry in txn_update_oldest results in a hang
* WT-2285       Enhance configure to set BUFFER_ALIGNMENT_DEFAULT to 4kb on Linux
* WT-2289       Fix a bug in btree search when doing a fast key check
* WT-2291       Random cursor walk inefficient in skip list only trees
* WT-2295       WT_SESSION.create does a full-scan of the main table
* WT-2296       Improve log algorithm for sync/flush settings
* WT-2297       Fix off-by-one error in Huffman config file parsing
* WT-2299       Clean up layering violation between btree and block manager code
* WT-2307       Fix a bug where internal page splits can corrupt cursor iteration
* WT-2308       Add support for custom extractor for ref_cursors in join cursor
* WT-2311       Add support for UltraSparc platform
* WT-2312       Fix a bug where re-creating a deleted column-store page can corrupt the in-memory tree
* WT-2313       Fix a bug in the sweep server
* WT-2314       Update page-swap error handling so that it is consistent
* WT-2316       Fix a bug in WT_CURSOR.prev where it could return keys out-of-order
* WT-2318       Enhance condition wait implementation to use less CPU on idle databases
* WT-2321       Fix a race between eviction and worker threads on the eviction queue
* WT-2322       Fix a bug in read-uncommitted join cursors where using Bloom filters is unsafe
* WT-2328       Update schema drop does to use the block manager interface for file removal
* WT-2331       Checking of search result for reference cursors before join
* WT-2332       Fix a bug in logging write-no-sync mode
* WT-2335       Fix a bug where parsing an invalid configuration string could segfault
* WT-2338       Disable using pre-allocated log files when a backup cursor is open
* WT-2339       Fix a bug in rebalance that caused database verification failure
* WT-2340       Add logging guarantee assertions
* WT-2345       Avoid creating tiny pages on disk when evicting small pages from cache
* WT-2346       Enhance checkpoint implementation so the schema lock is not held during I/O
* WT-2347       Fix some schema format edge cases in Java API
* WT-2352       Allow build and test without requiring lz4

src/mongo/gotools/mongooplog/main/mongooplog.go
// Main package for the mongooplog tool.
package main

import (
	"github.com/mongodb/mongo-tools/common/db"
	"github.com/mongodb/mongo-tools/common/log"
	"github.com/mongodb/mongo-tools/common/options"
	"github.com/mongodb/mongo-tools/common/signals"
	"github.com/mongodb/mongo-tools/common/util"
	"github.com/mongodb/mongo-tools/mongooplog"
	"os"
)

func main() {
	// initialize command line options
	opts := options.New("mongooplog", mongooplog.Usage,
		options.EnabledOptions{Auth: true, Connection: true, Namespace: false})

	// add the mongooplog-specific options
	sourceOpts := &mongooplog.SourceOptions{}
	opts.AddOptions(sourceOpts)

	log.Logvf(log.Always, "warning: mongooplog is deprecated, and will be removed completely in a future release")

	// parse the command line options
	args, err := opts.Parse()
	if err != nil {
		log.Logvf(log.Always, "error parsing command line options: %v", err)
		log.Logvf(log.Always, "try 'mongooplog --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	if len(args) != 0 {
		log.Logvf(log.Always, "positional arguments not allowed: %v", args)
		log.Logvf(log.Always, "try 'mongooplog --help' for more information")
		os.Exit(util.ExitBadOptions)
	}

	// print help, if specified
	if opts.PrintHelp(false) {
		return
	}

	// print version, if specified
	if opts.PrintVersion() {
		return
	}

	// init logger
	log.SetVerbosity(opts.Verbosity)
	signals.Handle()

	// connect directly, unless a replica set name is explicitly specified
	_, setName := util.ParseConnectionString(opts.Host)
	opts.Direct = (setName == "")
	opts.ReplicaSetName = setName

	// validate the mongooplog options
	if sourceOpts.From == "" {
		log.Logvf(log.Always, "command line error: need to specify --from")
		os.Exit(util.ExitBadOptions)
	}

	// create a session provider for the destination server
	sessionProviderTo, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to destination host: %v", err)
		os.Exit(util.ExitError)
	}

	// create a session provider for the source server
	opts.Connection.Host = sourceOpts.From
	opts.Connection.Port = ""
	sessionProviderFrom, err := db.NewSessionProvider(*opts)
	if err != nil {
		log.Logvf(log.Always, "error connecting to source host: %v", err)
		os.Exit(util.ExitError)
	}

	// initialize mongooplog
	oplog := mongooplog.MongoOplog{
		ToolOptions:         opts,
		SourceOptions:       sourceOpts,
		SessionProviderFrom: sessionProviderFrom,
		SessionProviderTo:   sessionProviderTo,
	}

	// kick it off
	if err := oplog.Run(); err != nil {
		log.Logvf(log.Always, "error: %v", err)
		os.Exit(util.ExitError)
	}

}
